python/ray/serve/tests/test_deploy_2.py

```diff
@@ -379,5 +379,45 @@ def check_num_waiters(target: int):
         print(time.time(), f"Confirmed number of replicas are at {i+1}.")
 
 
+def test_unallocated_replica_shutdown(serve_instance):
+    """Test that unallocated replicas are stopped immediately without waiting for graceful shutdown.
+
+    When a replica is in PENDING_ALLOCATION state (due to requiring resources that don't exist),
+    stopping it should be immediate and not wait for the graceful shutdown timeout.
+    """
+
+    @serve.deployment(
+        ray_actor_options={"resources": {"custom": 1}},  # Require non-existent resource
+        graceful_shutdown_timeout_s=10,
+    )
+    class CustomResourceDeployment:
+        def __init__(self):
+            self.ready = False
+
+        def __call__(self):
+            return "ready"
+
+    serve.run(CustomResourceDeployment.bind())
+    wait_for_condition(
+        check_deployment_status,
+        name="CustomResourceDeployment",
+        expected_status=DeploymentStatus.UNHEALTHY,
+    )
+
+    check_num_replicas_eq("CustomResourceDeployment", 1)
+
+    start = time.time()
+    serve.delete("CustomResourceDeployment")
+    deletion_time = time.time() - start
+    # deletion_time should be under the graceful shutdown timeout
+    assert deletion_time < 5, f"Deletion took {deletion_time}s, expected < 5s"
```

Comment:
Sometimes the CI can be unexpectedly slow because multiple tests are running at the same time under the hood. Could we do the following instead to conservatively avoid flaky CI:
1. Change `graceful_shutdown_timeout_s` to something very large like 1000
2. Run `serve.delete` with `blocking` set to false; instead, use `wait_for_condition` explained in (3)
3. Use `wait_for_condition` to check on whether the replica actor was killed (you can use `list_actors`, there are examples in this test file), with a timeout of 10 or 15 seconds.